### PR TITLE
Prepare Release 0.5.0 / Update to Substrate 3.0

### LINKED ISSUE #241: update deps to substrate 3.0.0

### PR DESCRIPTION
**Pull request by insipx**
- Bring changelog up to date with current changes
- incorporate changes from substrate 3.0 (Mostly lots of error type refactorings)
- bump parity-scale-codec from to 2.0 from 1.3
- bump parity-util-mem from 0.8 to 0.9
- bump kvdb from 0.8 to 0.9
- bump kvdb-rocksdb from 0.10 to 0.11

closes #241

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,29 +15,39 @@
...
   - [Changed] Archive now accepts just a postgres URL instead of a postgres URL split into its parts. This should
-  make configuring the archive more straightforward. Takes from environment variable `DATABASE_URL` if not passed to the 
+  make configuring the archive more straightforward. Takes from environment variable `DATABASE_URL` if not passed to the
   archive directly
 - [Added] Archive now reads the `CHAIN_DATA_DB` environment variable if the path to the backend chain database is not passed directly.
-- [Changed] `drive` changed from sync to async function
 - [Removed] Archive no longer needs an RPC url to function
+- [Changed] Printing out of Postgres URL in plain text
+- [Added] Maximum amount of blocks loaded at once can be configured with `max_block_load`
+- [Added] Two options to configuration file for State Tracing
```
Comment (dvdplm) [inline CHANGELOG.md:24]:
```suggestion
- [Added] Two new options to the configuration relative to state tracing:
```

I think we should explain what state tracing actually is somewhere. It's the big feature of this release, so worth spending some text on it imo. (A link to the wiki might suffice here)

File: CHANGELOG.md [hunk 2]
```diff
@@ -12,29 +15,39 @@
...
+  - `targets` for specifying runtime targets to trace in WASM
+  - `folder` where WASM blobs with tracing enabled are kept.
+- [Changed] Archive config is now separated into sections for readability. Migration can be achieved by looking at the new `archive.conf` in `polkadot-archive` or `node-template-archive` folders.
```
Comment (dvdplm) [inline CHANGELOG.md:27]:
```suggestion
- [Changed] Archive config is now separated into sections for readability. Migration is manual but looking at the new `archive.conf` in `polkadot-archive` or `node-template-archive` folders should help.
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -12,29 +15,39 @@
...
 
 ### Internal Changes
-- [QoL] upgrade to SQLx 0.4.0
-- [perf] Overhaul of block indexing. Now uses a Iterator to only collect batches of blocks from the database, 
+- [Changed] Postgres Queries are now type checked
```
Comment (dvdplm) [inline CHANGELOG.md:30]:
```suggestion
- [Changed] Postgres SQL queries are now type checked
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -12,29 +15,39 @@
...
+- [QoL] Refactor file layout to `substrate-archive` and `substrate-archive-backend`.
+- [perf] Decouple Database actors
+- [QoL] upgrade to SQLx 0.5.0
+- [perf] Overhaul of block indexing. Now uses a Iterator to only collect batches of blocks from the database,
 taking advantage of sequential read-speeds. Gathering blocks by RPC is no longer done.
```
Comment (dvdplm) [inline CHANGELOG.md:35]:
```suggestion
taking advantage of the better reading performance of sequential data access. Gathering blocks by RPC is no longer done.
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -12,29 +15,39 @@
   - [Changed] no longer need to instantiate a client and manually pass it to the Archive
   - [Changed] rename `run_with` to `drive`
   - [Changed] Archive now accepts just a postgres URL instead of a postgres URL split into its parts. This should
-  make configuring the archive more straightforward. Takes from environment variable `DATABASE_URL` if not passed to the 
+  make configuring the archive more straightforward. Takes from environment variable `DATABASE_URL` if not passed to the
   archive directly
 - [Added] Archive now reads the `CHAIN_DATA_DB` environment variable if the path to the backend chain database is not passed directly.
-- [Changed] `drive` changed from sync to async function
 - [Removed] Archive no longer needs an RPC url to function
+- [Changed] Printing out of Postgres URL in plain text
```
Comment (dvdplm) [inline CHANGELOG.md:22]:
```suggestion
- [Changed] Print out the Postgres URL at startup
```

…not including the password I hope.

File: CHANGELOG.md [hunk 6]
```diff
@@ -12,29 +15,39 @@
...
+- [Added] Maximum amount of blocks loaded at once can be configured with `max_block_load`
```
Comment (dvdplm) [inline CHANGELOG.md:23]:
```suggestion
- [Added] `max_block_load` to configure the maximum number of blocks loaded at once
```

**Review by dvdplm — CHANGES_REQUESTED** [review]
Looks good.

We need a clippy-pass, some edits and a description of what state tracing is and why it's cool.

File: substrate-archive-backend/src/runtime_version_cache.rs [hunk 7]
```diff
@@ -159,9 +160,11 @@ impl<B: BlockT> VersionRange<B> {
 }
 
 fn decode_version(version: &[u8]) -> Result<sp_version::RuntimeVersion> {
+	#[allow(clippy::redundant_slicing)]
```
Comment (insipx) [inline substrate-archive-backend/src/runtime_version_cache.rs:163]: I've tried a few variations of `::decode(&mut &version[..])` and the one clippy accepts is to pass in a `mut version`, ultimately decoding `&mut version` (without [..]).

However, the version clippy accepts leads to a runtime error where we can no longer decode the RuntimeVersion. I believe it has to do with this: https://github.com/rust-lang/rust-clippy/issues/6968

File: CHANGELOG.md (hunk 1 above)
Comment (insipx) [reply CHANGELOG.md]: Added link to https://github.com/paritytech/substrate-archive/wiki/6.)-State-Tracing-&-Balance-Reconciliation

File: substrate-archive-backend/src/runtime_version_cache.rs (hunk 7 above)
Comment (Jarcho) [reply substrate-archive-backend/src/runtime_version_cache.rs:164]: You can just reborrow the slice as that's what clippy will be suggesting after it's fixed. i.e. `&mut &*version`

File: substrate-archive-backend/src/database.rs [hunk 8]
```diff
@@ -129,35 +129,15 @@ impl ReadOnlyDB for SecondaryRocksDB {
 	}
 }
 
-type DBError = std::result::Result<(), sp_database::error::DatabaseError>;
+type DbError = std::result::Result<(), sp_database::error::DatabaseError>;
 //TODO: Remove panics with a warning that database has not been written to / is read-only
```
Comment (mattrutherford) [inline substrate-archive-backend/src/database.rs:133]: Can this TODO be taken care of in this PR?

File: substrate-archive-backend/src/read_only_backend/main_backend.rs [hunk 9]
```diff
@@ -121,4 +121,9 @@ impl<Block: BlockT, D: ReadOnlyDB + 'static> Backend<Block> for ReadOnlyBackend<
 	fn get_import_lock(&self) -> &parking_lot::RwLock<()> {
 		panic!("No lock exists for read only backend!")
```
Comment (mattrutherford) [inline substrate-archive-backend/src/read_only_backend/main_backend.rs:122]: Maybe change this to a warn too, if that's the strategy.

**Review by mattrutherford — APPROVED** [review]
LGTM - just a couple of places where there's some inconsistency in trait impls that panic, where most warn instead.

**Review by dvdplm — APPROVED** [review]

**Pull request merged**